pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -0,0 +1,181 @@
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	validatingWebhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())
+	return []admissionregistrationv1.ValidatingWebhook{*validatingWebhook}
+}
+
+// Admitters returns the admitter objects used to validate secrets.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct{}
+
+// Admit handles the webhook admission request sent to this webhook.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("project Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	oldProject, newProject, err := objectsv3.ProjectOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get old and new projects from request: %w", err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		return a.admitDelete(oldProject)
+	}
+	return a.admitCreateOrUpdate(oldProject, newProject)
+}
+
+func (a *admitter) admitDelete(project *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	if project.Labels[systemProjectLabel] == "true" {
+		return admission.ResponseBadRequest("System Project cannot be deleted"), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func (a *admitter) admitCreateOrUpdate(oldProject, newProject *v3.Project) (*admissionv1.AdmissionResponse, error) {
+	projectQuota := newProject.Spec.ResourceQuota
+	nsQuota := newProject.Spec.NamespaceDefaultResourceQuota
+	if projectQuota == nil && nsQuota == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	fieldErr, err := checkQuotaFields(projectQuota, nsQuota)
+	if err != nil {
+		return nil, fmt.Errorf("error checking project fields: %w", err)
+	}
+	if fieldErr != nil {
+		return admission.ResponseBadRequest(fieldErr.Error()), nil
+	}
+	fieldErr, err = a.checkQuotaValues(&nsQuota.Limit, &projectQuota.Limit, oldProject)
+	if err != nil {
+		return nil, fmt.Errorf("error checking quota values: %w", err)
+	}
+	if fieldErr != nil {
+		return admission.ResponseBadRequest(fieldErr.Error()), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func checkQuotaFields(projectQuota *v3.ProjectResourceQuota, nsQuota *v3.NamespaceResourceQuota) (*field.Error, error) {
+
+	if projectQuota == nil && nsQuota != nil {
+		return field.Required(projectSpecFieldPath.Child(projectQuotaField), fmt.Sprintf("required when %s is set", namespaceQuotaField)), nil
+	}
+	if projectQuota != nil && nsQuota == nil {
+		return field.Required(projectSpecFieldPath.Child(namespaceQuotaField), fmt.Sprintf("required when %s is set", projectQuotaField)), nil
+	}
+
+	projectQuotaLimitMap, err := convert.EncodeToMap(projectQuota.Limit)
```

Comment:
A field error would be for something the user could control. In this case the only thing that could cause the encoding to fail would be passing it a bad data structure, but since we control the data structure that would be a programmatic error, not the user's fault.

Reply:
Right, but what's the main reason for which EncodeToMap might return an error? Mostly bad data, if I understand correctly. Although, using EncodeToMap is our implementation detail, and we might move away from it or add something else, so then it's a server error and not the user's fault...

Reply:
What's an example of bad data a user might be able to pass that would be accepted by the apiserver but fail to encode here? I am unable to find an example: https://go.dev/play/p/TWSRiAjif-j

Reply:
The, I wonder, if we don't even need to handle that error if we don't foresee one being returned?

Reply:
> The, I wonder, if we don't even need to handle that error if we don't foresee one being returned?

I don't agree with this viewpoint - just because we can't figure out a case now doesn't mean it can't happen. Generally speaking, IMO production code should never ignore an error.